mlir/include/mlir/Interfaces/ValueBoundsOpInterface.h

```diff
@@ -113,8 +113,9 @@ class ValueBoundsConstraintSet {
   ///
   /// The first parameter of the function is the shaped value/index-typed
   /// value. The second parameter is the dimension in case of a shaped value.
-  using StopConditionFn =
-      function_ref<bool(Value, std::optional<int64_t> /*dim*/)>;
+  /// The third parameter is this constraint set.
+  using StopConditionFn = std::function<bool(
+      Value, std::optional<int64_t> /*dim*/, ValueBoundsConstraintSet &cstr)>;
```

Comment:
This comment is not directly related to the changes in this PR, but It's not immediately clear to me if which value indicates stop/continuation. I like the `walk` function much more which exposes `WalkResult::interrup()` and `WalkRedult::advance()`, but I think that changing to `LogicalResult` and renaming this could also help.
Would be cool if we could revisit this separately.

Reply:
Good point, I am also thinking of splitting up `stopCondition` into two functions, so that IR traversal boundary and values that are projected out can be specified separately. We had a use case for that recently.

Reply:
I'm going to prepare a follow-up PR for this, after my chain of PRs has landed.